mrp_production_batch/models/mrp_production.py

```diff
@@ -0,0 +1,108 @@
+# Copyright NuoBiT Solutions - Kilian Niubo <kniubo@nuobit.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class MrpProduction(models.Model):
+    _inherit = "mrp.production"
+
+    production_batch_id = fields.Many2one(
+        comodel_name="mrp.production.batch",
+        string="Production Batch",
+        ondelete="restrict",
+        domain="[('state', '!=', 'done')]",
+    )
+
+    def _check_production_to_batch_consistency(self, mrp_productions):
+        if not mrp_productions:
+            raise ValidationError(_("No productions selected"))
+        if mrp_productions.production_batch_id:
+            raise ValidationError(
+                _("Some of the selected productions already have a batch")
+            )
+        if {"done", "cancel"} & set(mrp_productions.mapped("state")):
+            raise ValidationError(
+                _("Some of the selected productions are already done or cancelled")
+            )
+        if len(mrp_productions.mapped("picking_type_id")) > 1:
```

Comment:
Not necessary to used mapped on a recordset, only on simple types
` if len(mrp_productions.picking_type_id") > 1:`